crates/wasmtime/src/runtime/rr/hooks/component_hooks.rs

```diff
@@ -0,0 +1,321 @@
+) -> Result<()>
+where
+    F: FnOnce(&mut LowerContext<'_, T>, InterfaceType, usize) -> Result<()>,
+{
+    #[cfg(feature = "rr-component")]
+    cx.store
+        .0
+        .record_event_validation(|| LowerMemoryEntryEvent { ty, offset })?;
+    let store_result = lower_store(cx, ty, offset);
+    #[cfg(feature = "rr-component")]
+    cx.store
+        .0
+        .record_event(|| LowerMemoryReturnEvent(ResultEvent::from_anyhow_result(&store_result)))?;
+    store_result
+}
+
+/// Record hook wrapping a flat lowering call of component types
+#[inline]
+pub fn record_lower_flat<F, T>(
+    lower: F,
+    cx: &mut LowerContext<'_, T>,
+    ty: InterfaceType,
+) -> Result<()>
+where
+    F: FnOnce(&mut LowerContext<'_, T>, InterfaceType) -> Result<()>,
+{
+    #[cfg(feature = "rr-component")]
+    cx.store
+        .0
+        .record_event_validation(|| LowerFlatEntryEvent { ty })?;
+    let lower_result = lower(cx, ty);
+    #[cfg(feature = "rr-component")]
+    cx.store
+        .0
+        .record_event(|| LowerFlatReturnEvent(ResultEvent::from_anyhow_result(&lower_result)))?;
+    lower_result
+}
+
+#[cfg(feature = "rr-component")]
+#[inline(always)]
+fn create_host_func_entry_event(
+    args: &mut [MaybeUninit<ValRaw>],
+    types: &Arc<ComponentTypes>,
+    param_tys: &InterfaceType,
+) -> HostFuncEntryEvent {
+    let flat_params = types.flat_types_storage_or_pointer(param_tys, MAX_FLAT_PARAMS);
+    HostFuncEntryEvent {
+        args: RRFuncArgVals::from_flat_storage(args, flat_params),
+    }
+}
+
+/// Same as [`ConstMemorySliceCell`] except allows for dynamically sized slices.
+///
+/// Prefer the above for efficiency if slice size is known statically.
+///
+/// **Note**: The correct operation of this type relies of several invariants.
+/// See [`ConstMemorySliceCell`] for detailed description on the role
+/// of these types.
+pub struct MemorySliceCell<'a> {
+    pub bytes: &'a mut [u8],
+    #[cfg(feature = "rr-component")]
+    pub offset: usize,
+    #[cfg(feature = "rr-component")]
+    pub recorder: Option<&'a mut RecordBuffer>,
+}
+impl<'a> Deref for MemorySliceCell<'a> {
+    type Target = [u8];
+    fn deref(&self) -> &Self::Target {
+        self.bytes
+    }
+}
+impl DerefMut for MemorySliceCell<'_> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.bytes
+    }
+}
+impl Drop for MemorySliceCell<'_> {
+    /// Drop serves as a recording hook for stores to the memory slice
+    fn drop(&mut self) {
+        #[cfg(feature = "rr-component")]
+        if let Some(buf) = &mut self.recorder {
+            buf.record_event(|| MemorySliceWriteEvent {
+                offset: self.offset,
+                bytes: self.bytes.to_vec(),
+            })
+            .unwrap();
+        }
+    }
+}
+
+/// Zero-cost encapsulation type for a statically sized slice of mutable memory
+///
+/// # Purpose and Usage (Read Carefully!)
+///
+/// This type (and its dynamic counterpart [`MemorySliceCell`]) are critical to
```

Comment:
naming nit: usually `Const` means compile-time-constant in Rust. Let's use `MemorySliceCell` and `MemorySliceCellMut`?

Also, `Cell` means something specific (atomic unit of interior mutability); let's use `Ref` maybe? So `MemorySliceRef` and `MemorySliceRefMut`?